File: dynamic_listings/lua/default/design_patterns/proxy.lua
-- This class represents a user, with their permissions
User = {}

function User:can_do_thing()
    -- Changes according to user permissions
    -- ...
end

-- This class represents a web request
Request = {}

function Request:new(user)
    o = {
        user = user
    }
    setmetatable(o, self)
    self.__index = self
    return o
end

-- Represents a call to a web page
WebPage = {}

function WebPage:new(o)
    o = o or {}
    setmetatable(o, self)
    self.__index = self
    return o
end

function WebPage:get(request)
    -- This function gets implemented in the concrete classes;
end

-- Represents an authentication proxy for a web page
WebPageProxy = WebPage:new()

function WebPageProxy:get(request)
    -- Get the requesting user
    local requesting_user = request.user
    if requesting_user.can_do_thing(request) then
        return WebPage:get(request)
    else
        return HttpResponse:new("This user cannot access this page")
    end
end

function main()
    local user = User:new();
    -- The following line is valid due to polimorphism
    local page = WebPageProxy:new();
    -- If the user has permission the page will come through.
    page.get(Request:new(user));
end

File: dynamic_listings/javascript/default/design_patterns/proxy.js
class User{
    // This class represents a user, with their permissions
    // ...
    can_do_thing(){
        // Changes according to user permissions
        // ...
    }
}

class Request{
    // This class represents a web request
    constructor(user){
        this.user = user;
    }
}

class WebPage{
    /*
     * Represents a call to a web page
     */
    // This function gets implemented in the concrete classes;
    get(request){
        throw "Cannot use get() from abstract class"
    }
}

class WebPageProxy extends WebPage{
    // Represents an authentication proxy for a web page
    get(request){
        // Get the requesting user
        requesting_user = request.user;
        if (requesting_user.can_do_thing()){
            return super.get(request);
        }else{
            return HttpResponse("This user cannot access this page");
        }
    }
}

let user = User();
// The following line is valid due to polimorphism
let page = WebPageProxy();
// If the user has permission the page will come through.
page.get(Request(user));
